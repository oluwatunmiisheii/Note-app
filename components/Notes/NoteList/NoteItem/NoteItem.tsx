import { useNavigation } from "@react-navigation/native";
import { View, Text, StyleSheet, Pressable } from "react-native";
import { Note } from "../../../../models/note.model";
import { getFormattedDate } from "../../../../utils/date";
import {
  Gesture,
  GestureDetector,
  gestureHandlerRootHOC,
} from "react-native-gesture-handler";

interface NoteItemProps {
  note: Note;
  showActions: (note: Note) => void;
}

export const NoteItem = gestureHandlerRootHOC(
  ({ note, showActions }: NoteItemProps) => {
    const navigation = useNavigation();

    const goToDetailsScreen = () => {
      navigation.navigate("NoteDetailsScreen", { noteId: note.id });
    };

    const longPressGesture = Gesture.LongPress().onEnd((event, success) => {
      if (success) {
        showActions(note);
      }
    });

    return (
      <GestureDetector gesture={longPressGesture}>
        <Pressable style={styles.noteContainer} onPress={goToDetailsScreen}>
          <View style={styles.noteHeader}>
            <Text
              style={styles.noteTitle}
              numberOfLines={1}
              ellipsizeMode="tail"
            >
              {note.title}
            </Text>
            <Text>{getFormattedDate(new Date(note.date))}</Text>
          </View>
          <Text numberOfLines={2} ellipsizeMode="tail">
            {note.description}
          </Text>
        </Pressable>
      </GestureDetector>
    );
  }
);

const styles = StyleSheet.create({
  noteContainer: {
    width: "100%",
    backgroundColor: "#FDE69A",
    borderColor: "#FDE69A",
    borderWidth: 1,
    borderRadius: 6,
    marginBottom: 20,
    padding: 12,
    justifyContent: "space-between",
    elevation: 3,
    shadowColor: "#000",
    shadowRadius: 2,
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.1,
  },
  noteHeader: {
    width: "100%",
    justifyContent: "space-between",
    flexDirection: "row",
  },
  noteTitle: {
    fontSize: 18,
    fontWeight: "700",
    maxWidth: "76%",
    flex: 1,
    marginBottom: 5,
  },
});
