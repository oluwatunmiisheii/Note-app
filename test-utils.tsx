/* eslint-disable no-undef, import/no-extraneous-dependencies */
import * as React from "react";
import { NavigationContainer } from "@react-navigation/native";
import { render } from "@testing-library/react-native";
import renderer from "react-test-renderer";

/**
 * Render given JSX inside Navigation container.
 * This should be used for rendering whole navigators as used by real app.
 */
export function renderNavigator(ui: React.ReactElement) {
  return render(<NavigationContainer>{ui}</NavigationContainer>);
}

export function createNavigator(ui: React.ReactElement) {
  const tree = renderer
    .create(<NavigationContainer>{ui}</NavigationContainer>)
    .toJSON();
  return { tree };
}
