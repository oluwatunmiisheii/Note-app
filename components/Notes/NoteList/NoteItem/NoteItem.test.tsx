import { screen, fireEvent } from "@testing-library/react-native";
import { getNote } from "../../../../services/note/__mock_data__/notes.mock";
import { NoteItem } from "./NoteItem";
import { createNavigator, renderNavigator } from "../../../../test-utils";

const note = getNote();
const showActions = jest.fn();

const defaultProps: React.ComponentProps<typeof NoteItem> = {
  note,
  showActions,
};

const create = (props: Partial<typeof defaultProps> = defaultProps) => {
  const { tree } = createNavigator(
    <NoteItem {...{ ...defaultProps, ...props }} />
  );
  return { props, tree };
};

const setup = (props: Partial<typeof defaultProps> = defaultProps) => {
  const utils = renderNavigator(
    <NoteItem {...{ ...defaultProps, ...props }} />
  );
  return { props, ...utils };
};

const mockNavigate = jest.fn();

// mock navigation
jest.mock("@react-navigation/native", () => ({
  ...jest.requireActual("@react-navigation/native"),
  useNavigation: jest.fn(() => ({
    navigate: mockNavigate,
  })),
}));

describe(`<${NoteItem.name}>`, () => {
  test("renders correctly", () => {
    const { tree } = create();
    expect(tree).toMatchSnapshot();
  });

  test("renders note title", () => {
    setup();
    expect(screen.getByText(note.title)).toBeOnTheScreen();
  });

  test("renders note description", () => {
    setup();
    expect(screen.getByText(note.description)).toBeOnTheScreen();
  });

  test("should redirect to note details page on tap", async () => {
    setup();
    const trigger = screen.getByText(note.title);
    fireEvent(trigger, "press");
    expect(mockNavigate).toBeCalledTimes(1);
  });
});
