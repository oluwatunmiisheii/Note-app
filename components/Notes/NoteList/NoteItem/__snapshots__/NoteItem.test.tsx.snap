// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Wrapper> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#FDE69A",
        "borderColor": "#FDE69A",
        "borderRadius": 6,
        "borderWidth": 1,
        "elevation": 3,
        "justifyContent": "space-between",
        "marginBottom": 20,
        "padding": 12,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 2,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "flex": 1,
            "fontSize": 18,
            "fontWeight": "700",
            "marginBottom": 5,
            "maxWidth": "76%",
          }
        }
      >
        Note title
      </Text>
      <Text>
        2023-2-15
      </Text>
    </View>
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
    >
      Note description
    </Text>
  </View>
</View>
`;
